torch/_library/abstract_impl.py

```diff
@@ -0,0 +1,197 @@
+                f"impl_abstract(...): the operator {self.qualname} "
+                f"already has an abstract impl registered at "
+                f"{self.kernel.source}."
+            )
+        if torch._C._dispatch_has_kernel_for_dispatch_key(self.qualname, "Meta"):
+            raise RuntimeError(
+                f"impl_abstract(...): the operator {self.qualname} "
+                f"already has an DispatchKey::Meta implementation via a "
+                f"pre-existing torch.library or TORCH_LIBRARY registration. "
+                f"Please either remove that registration or don't call "
+                f"impl_abstract."
+            )
+
+        if torch._C._dispatch_has_kernel_for_dispatch_key(
+            self.qualname, "CompositeImplicitAutograd"
+        ):
+            raise RuntimeError(
+                f"impl_abstract(...): the operator {self.qualname} "
+                f"already has an implementation for this device type via a "
+                f"pre-existing registration to "
+                f"DispatchKey::CompositeImplicitAutograd."
+                f"CompositeImplicitAutograd operators do not need an abstract "
+                f"impl; "
+                f"instead, the operator will decompose into its constituents "
+                f"and those "
+                f"can have abstract impls defined on them."
+            )
+
+        # Store the kernel in this holder
+        self.kernel = Kernel(func, source)
+
+        # Also register the abstract impl to Meta key
+        if self.lib is None:
+            ns = self.qualname.split("::")[0]
+            self.lib = torch.library.Library(ns, "FRAGMENT")
+        meta_kernel = construct_meta_kernel(self.qualname, self)
+        self.lib.impl(self.qualname, meta_kernel, "Meta")
+
+        def deregister_abstract_impl():
+            if self.lib:
+                self.lib._destroy()
+                self.lib = None
+            self.kernel = None
+
+        return RegistrationHandle(deregister_abstract_impl)
+
+
+def construct_meta_kernel(
+    qualname: str, abstract_impl_holder: AbstractImplHolder
+) -> Callable:
+    assert abstract_impl_holder.kernel is not None
+
+    @functools.wraps(abstract_impl_holder.kernel.func)
+    def meta_kernel(*args, **kwargs):
+        assert abstract_impl_holder.kernel is not None
+        source = abstract_impl_holder.kernel.source
+
+        def error_on_ctx():
+            raise RuntimeError(
+                f"Attempted to call get_ctx() for the meta implementation "
+                f"for {qualname}."
+                f"You have presumably called get_ctx() because the operator "
+                f"has a data-dependent output shape; if so, there is no "
+                f"such meta implementation and this error is the correct "
+                f"behavior. Otherwise, please remove the call to get_ctx() "
+                f"in the implementation registered with impl_abstract "
+                f"at {source}"
```

Comment:
huh, is this a thing lol

Reply:
Meta formulas can't use the get_ctx() because there isn't a shape env. There are a couple of things we can do here:
- get_ctx() for meta errors out (what is here)
- get_ctx() for meta returns None
- get_ctx().new_dynamic_size() errors.

and we went with the first one

Reply:
I guess my point is, I would only do get_ctx() if I was about to do data dependent output. So the case where I can just "remove get_ctx()" seems unlikely, and not worth calling out

Reply:
Fair enough